Upgrade iOS notifications to User Notifications Framework

Szymon20000 opened the pull request:
# Why

We've recently dropped support for iOS 9 and it has unblocked us to move to `UNNotificationFramework`.

# How

- removed old methods from `AppDelegate`
- removed `EXLocalNotificationRequester` in favor of `EXUserNotificationRequester` (they both serve the same purpose — asking for notifications permission)
- initial notification isn't passed down from `launchOptions`, but is asynchronously fetched by `EXReactAppManager` from `EXUserNotificationManager` which is a delegate of `UNNotificationCenter`. It receives all the notifications the system wants us to receive, that is of `willPresentNotification` and `didReceiveNotificationResponse`.
- `EXUserNotificationManager` is now a kernel service. Always use it through `EXKernelServiceRegistry`.
- created an `EXPendingNotification` class encapsulating notification data. It has become the go-to class to store and pass notification around in kernel. It discerns between push and local notification by trigger class.
- added ability to create notification categories — `Notifications.createCategoryIOSAsync`. The categories IDs are prefixed with experience ID in scoped context (Expo Client), but are used as-provided in standalone and ejected projects
- `notification` object (the one in `this.props.exp.notification` and the one that is passed to notifications listener) got two new properties — `actionId` and `userText`. `actionId` denotes an action identifier triggered by the user (empty if the user just tapped the notification). `userText` is the text provided by the user if the action allowed text input.
- `UNNotificationFramework` doesn't allow us to create a repeating notification starting some day in the future (it's not possible to easily schedule a notification that will trigger every minute, starting in 3 hours). We have to either remove the ability to use `repeats` on iOS or use `UILocalNotification` for now. [Discussion pending.](
https://github.com/expo/expo/pull/2316#discussion_r233382185)
- To ensure thread-safety, interaction with `UNNotificationCenter` always goes through a proxy. In ExpoKit it's `EXUserNotificationCenter`, which defines its own queue to which it dispatches all the calls, in vanilla apps it's `EXUserNotificationCenterProxy`. It's an internal module (Expo universal module architecture) implementing `EXUserNotificationCenterProxyInterface`.
- `EXPermissions` now implements a new interface, `EXPermissionsModule` which exports a `-methodQueue` method. It allows us to safely ask `UNNotificationCenter` for authorization settings in `EXUserNotificationRequester`. Due to its completion-handlery nature we have to sprinkle `EXUserNotificationRequester` with semaphores magic to mimic synchronous behavior of the previous API.

# Test plan

_(most of the unchecked ones were working ok iirc, but I want to confirm that again before checking the box)_

- [x] presented notification doesn't show,
- [x] presented notification triggers `Notifications` listener
- [x] scheduled notification doesn't show if application is active
- [x] scheduled notification shows if application is backgrounded
- [x] scheduled notification shows if application is dead
- [x] scheduled and remote notifications with custom categories use those categories
- [x] scheduled, tapped notification triggers `Notifications` listener
- [x] overwriting categories works properly
- [x] deleting categories works properly
- [x] Expo Client switches to a proper experience when a notification leading to a closed experience is tapped (app is just backgrounded)
- [x] Expo Client doesn't switch to any experience when a notification is tapped while the app is inactive
- [x] canceling a scheduled notification by its ID works
- [x] canceling all scheduled notifications works
- [x] experiences don't get initial notifications in Expo Client (although it could be implemented)
- [x] standalone apps receive initial notification in `this.props.exp.notification`
- [x] asking for user notifications permission resolves in both cases
- [x] asking for remote notifications permission resolves in both cases
- [x] a legacy notification uses custom category
- [ ] update documentation
- [x] update Jest mocks
- [x] `repeats` — remove vs use legacy class discussion (added legacy method with a warning)
- [x] repeat all these tests on iOS < 12 (done on iOS 10)
- [x] repeat all these tests in a standalone app

sjchmiela commented on ios/Exponent/Kernel/Services/Notifications/EXRemoteNotificationManager.m:
```diff
@@ -208,26 +194,31 @@ - (void)_synchronizeCurrentAPNSToken:(NSData *)currentToken
   // When the user has not granted permission to display any type of notification, iOS doesn't
   // invoke the delegate methods and registering for remote notifications will never complete
   //
   // TODO: Switch this to use UNNotificationSettings for iOS 10+. Note that UNNotificationCenter
```
We could finally remove this comment 😅

sjchmiela commented on ios/Exponent/Kernel/Services/Notifications/EXRemoteNotificationManager.m:
```diff
@@ -208,26 +194,31 @@ - (void)_synchronizeCurrentAPNSToken:(NSData *)currentToken
-  UIUserNotificationSettings * settings = RCTSharedApplication().currentUserNotificationSettings;
-  return settings.types != UIUserNotificationTypeNone;
+  [[UNUserNotificationCenter currentNotificationCenter] getNotificationSettingsWithCompletionHandler:^(UNNotificationSettings * _Nonnull settings) {
+    if (settings.authorizationStatus == UNAuthorizationStatusAuthorized) {
```
`handler(settings.authorizationStatus == UNAuthorizationStatusAuthorized)` would be a bit more concise 🙂

sjchmiela commented on ios/Exponent/Supporting/Info.plist:
```diff
@@ -49,7 +49,7 @@
 	<dict>
 		<key>APIKey</key>
-		<string>a25fafe9f9edee11a9882b32e0cd7a26df6e2c42</string>
+		<string>81130e95ea13cd7ed9a4f455e96214902c721c99</string>
```
Why is it changed? 😱

sjchmiela commented on ios/Exponent/Versioned/Core/Api/EXNotifications.m:
```diff
@@ -92,41 +93,116 @@ - (instancetype)initWithExperienceId:(NSString *)experienceId kernelServiceDeleg
 {
-  UILocalNotification *notification = [self _localNotificationFromPayload:payload];
 
+  UNMutableNotificationContent* content = [self _localNotificationFromPayload:payload];
```
💅 remove extra line above and add one below 🙂

sjchmiela commented on ios/Exponent/Versioned/Core/Api/EXNotifications.m:
```diff
@@ -92,41 +93,116 @@ - (instancetype)initWithExperienceId:(NSString *)experienceId kernelServiceDeleg
   [EXUtil performSynchronouslyOnMainThread:^{
-    [RCTSharedApplication() presentLocalNotificationNow:notification];
+    UNTimeIntervalNotificationTrigger* trigger = [UNTimeIntervalNotificationTrigger triggerWithTimeInterval:1 repeats:NO];
+    UNNotificationRequest* request = [UNNotificationRequest
```
> If the request does not contain a UNNotificationTrigger object, the notification is delivered right away.
> source: https://developer.apple.com/documentation/usernotifications/unusernotificationcenter/1649508-addnotificationrequest?language=objc

So we could not pass `trigger` and achieve something similar, right?

sjchmiela commented on ios/Exponent/Versioned/Core/Api/EXNotifications.m:
```diff
@@ -92,41 +93,116 @@ - (instancetype)initWithExperienceId:(NSString *)experienceId kernelServiceDeleg
+                                      requestWithIdentifier:content.userInfo[@"id"] content:content trigger:trigger];
+    [[UNUserNotificationCenter currentNotificationCenter] addNotificationRequest:request withCompletionHandler:^(NSError * _Nullable error) {
+      if (error != nil) {
+        NSLog(@"%@", error.localizedDescription);
```
Obsolete `NSLog`

sjchmiela commented on ios/Exponent/Versioned/Core/Api/EXNotifications.m:
```diff
@@ -92,41 +93,116 @@ - (instancetype)initWithExperienceId:(NSString *)experienceId kernelServiceDeleg
+    [[UNUserNotificationCenter currentNotificationCenter] addNotificationRequest:request withCompletionHandler:^(NSError * _Nullable error) {
+      if (error != nil) {
+        NSLog(@"%@", error.localizedDescription);
+        reject(@"Could not make notification request", error.localizedDescription, error);
```
First argument to `reject` should be error code, second — message and third — an error, right?

I think `E_NOTIF` would work better as the code and
`Could not add a notification request: ${error.description}` as the message.

sjchmiela commented on ios/Exponent/Versioned/Core/Api/EXNotifications.m:
```diff
@@ -92,41 +93,116 @@ - (instancetype)initWithExperienceId:(NSString *)experienceId kernelServiceDeleg
+}
+
+RCT_EXPORT_METHOD(addCategory: (NSString *) categoryId
+                  actions: (NSArray *) actions
```
💅 extra spaces

sjchmiela commented on ios/Exponent/Versioned/Core/Api/EXNotifications.m:
```diff
@@ -92,41 +93,116 @@ - (instancetype)initWithExperienceId:(NSString *)experienceId kernelServiceDeleg
+  
+  UNNotificationCategory * newCategory = [UNNotificationCategory categoryWithIdentifier:categoryId actions:actionsArray intentIdentifiers:@[] options:UNNotificationCategoryOptionNone];
+  
+  [[UNUserNotificationCenter currentNotificationCenter] setNotificationCategories:[NSSet setWithObject:newCategory]];
```
As we've agreed, `setNotificationCategories` probably overwrites other categories, so we have to first `getNotificationCategories` and then in the callback add this new category.

sjchmiela commented on ios/Exponent/Versioned/Core/Api/EXNotifications.m:
```diff
@@ -92,41 +93,116 @@ - (instancetype)initWithExperienceId:(NSString *)experienceId kernelServiceDeleg
+{
+  NSMutableArray<UNNotificationAction *> * actionsArray = [[NSMutableArray alloc] init];
+  
+  for( NSArray * action in actions) {
```
💅 formatting

sjchmiela commented on ios/Exponent/Versioned/Core/Api/EXNotifications.m:
```diff
@@ -92,41 +93,116 @@ - (instancetype)initWithExperienceId:(NSString *)experienceId kernelServiceDeleg
+  UNNotificationCategory * newCategory = [UNNotificationCategory categoryWithIdentifier:categoryId actions:actionsArray intentIdentifiers:@[] options:UNNotificationCategoryOptionNone];
+  
+  [[UNUserNotificationCenter currentNotificationCenter] setNotificationCategories:[NSSet setWithObject:newCategory]];
+  resolve(@"done");
```
`@"done"` isn't the best return value, `actionId` would work better or just `@(YES)`

sjchmiela commented on ios/Exponent/Versioned/Core/Api/EXNotifications.m:
```diff
@@ -92,41 +93,116 @@ - (instancetype)initWithExperienceId:(NSString *)experienceId kernelServiceDeleg
+                                                                                 textInputButtonTitle:action[3] textInputPlaceholder:action[4]];
+      [actionsArray addObject:newAction];
+    } else {
+      UNNotificationAction * newAction = [UNNotificationAction actionWithIdentifier:action[0] title:action[1] options:options];
```
We'll need to prefix `action[id]` with `experienceId` not to let experiences tamper with each others categories 🙂

sjchmiela commented on ios/Exponent/Versioned/Core/Api/EXNotifications.m:
```diff
@@ -141,13 +217,14 @@ - (instancetype)initWithExperienceId:(NSString *)experienceId kernelServiceDeleg
-    for (UILocalNotification *notification in [RCTSharedApplication() scheduledLocalNotifications]) {
-      if ([notification.userInfo[@"experienceId"] isEqualToString:self.experienceId]) {
-        [RCTSharedApplication() cancelLocalNotification:notification];
+  [[UNUserNotificationCenter currentNotificationCenter] getPendingNotificationRequestsWithCompletionHandler:
```
Doesn't it need to run on the main thread?

Szymon20000 replied:
"You may use the shared user notification center object simultaneously from any of your app's threads. The object processes requests serially in the order they were initiated."
https://developer.apple.com/documentation/usernotifications/unusernotificationcenter?language=objc

ide replied:
In iOS 11 the guidance was:

>Thread-Safety Information
>
>You may use the shared user notification center object from any thread of your app. However, you should use this object from only one thread at a time. Do not try to use it from multiple threads simultaneously. Always assign your delegate object to the shared user notification center's delegate property before using the object. It is a programmer error to assign a delegate after calling methods that might return information to that delegate.

So for backwards compatibility I think we may need to use a consistent queue everywhere. But unlike with UIApplication, we don't need to use the main queue.

Szymon20000 replied:
I just didn't notice that. I think that we can wrap unusernotificationcenter (only used methods) in a class and each method  will be pushing block to our queue.

sjchmiela commented on ios/Exponent/Versioned/Core/Api/EXNotifications.m:
```diff
@@ -92,41 +93,116 @@ - (instancetype)initWithExperienceId:(NSString *)experienceId kernelServiceDeleg
-        break;
-      }
-    }
+    [[UNUserNotificationCenter currentNotificationCenter] removePendingNotificationRequestsWithIdentifiers:@[uniqueId]];
```
Does it need to be run on the main thread?

sjchmiela commented on ios/Exponent/Versioned/Core/Api/EXNotifications.m:
```diff
@@ -178,29 +255,34 @@ - (instancetype)initWithExperienceId:(NSString *)experienceId kernelServiceDeleg
+  content.userInfo = @{
+   @"body": payload[@"data"],
+   @"experienceId": self.experienceId,
+   @"id": uniqueId
```
💅 formatting, copy and paste the whole dictionary

sjchmiela commented on ios/Exponent/Versioned/Core/Api/EXNotifications.m:
```diff
@@ -178,29 +255,34 @@ - (instancetype)initWithExperienceId:(NSString *)experienceId kernelServiceDeleg
-  RCTAssert((payload[@"data"] != nil), @"Attempted to send a local notification with no `data` property.");
-  UILocalNotification *localNotification = [UILocalNotification new];
-
+  
```
💅 extra newline

sjchmiela commented on ios/Exponent/Versioned/Core/Api/EXNotifications.m:
```diff
@@ -92,41 +93,116 @@ - (instancetype)initWithExperienceId:(NSString *)experienceId kernelServiceDeleg
-    [RCTSharedApplication() presentLocalNotificationNow:notification];
+    UNTimeIntervalNotificationTrigger* trigger = [UNTimeIntervalNotificationTrigger triggerWithTimeInterval:1 repeats:NO];
+    UNNotificationRequest* request = [UNNotificationRequest
+                                      requestWithIdentifier:content.userInfo[@"id"] content:content trigger:trigger];
```
I don't know if letting developers pass in `id` is a good idea… 🤔 Maybe we could generate it on the fly?

sjchmiela commented on ios/Exponent/Versioned/Core/Api/EXNotifications.m:
```diff
@@ -92,41 +93,116 @@ - (instancetype)initWithExperienceId:(NSString *)experienceId kernelServiceDeleg
+  NSDateComponents * date = [[NSDateComponents alloc] init];
+  NSArray * unites = @[@"day", @"month", @"year", @"weekday", @"quarter", @"leapMonth", @"nanosecond", @"era", @"weekdayOrdinal", @"weekOfMonth", @"weekOfYear", @"hour", @"second", @"minute", @"yearForWeekOfYear"];
+  for( NSString * unit in unites) {
+    if (options[unit]) [date setValue:(NSNumber *)options[unit] forKey:unit];
```
Maybe we should stick to supporting `repeat` https://docs.expo.io/versions/v30.0.0/sdk/notifications#repeat-optional-string----minute-hour-day?

sjchmiela reviewed (commented):
Thanks for the great work! I tried to notice all the little things, but if not – sorry!

Szymon20000 commented on apps/native-component-list/screens/UserNotifications.js:
```diff
@@ -0,0 +1,108 @@
+        data: { scheduledAt: new Date().getTime() },
+      },
+      {
+        'time-interval': 10,
```
That solution is obvious but we cannot support repeat option and that's why I added two new functions.

Szymon20000 commented on ios/Exponent/Kernel/Core/EXKernel.m:
```diff
@@ -26,7 +27,7 @@
 
 @interface EXKernel () <EXKernelAppRegistryDelegate>
-
+@property (atomic, strong) EXSendNotificationParams * sendNotificationParams;
```
There is no need for that. We are saving notification only when app wasn't running before. So there can be only one such notification.

Szymon20000 commented on ios/Client/EXRootViewController.m:
```diff
@@ -178,6 +178,10 @@ - (BOOL)isNuxFinished
 - (void)appDidFinishLoadingSuccessfully:(EXKernelAppRecord *)appRecord
 {
+  
+  if (appRecord == [EXKernel sharedInstance].appRegistry.homeAppRecord) {
```
Home is used only for finding last known url for the experienceId. We are just waiting until a listener in Home is accessible.

Szymon20000 commented on ios/Exponent/Kernel/Core/EXKernel.m:
```diff
@@ -139,10 +140,22 @@ - (void)sendUrl:(NSString *)urlString toAppRecord:(EXKernelAppRecord *)app
+- (void)runIfNotificationIsPresent
+{
+  if (_sendNotificationParams == nil) return;
+  [self sendNotification:_sendNotificationParams.dic
```
With a new framework, we cannot longer access local notification from launchOptions.

ide commented on ios/Exponent/Versioned/Core/Api/EXNotifications.m:
```diff
@@ -92,41 +93,124 @@ - (instancetype)initWithExperienceId:(NSString *)experienceId kernelServiceDeleg
+    }
+  }
+  
+  UNNotificationCategory * newCategory = [UNNotificationCategory categoryWithIdentifier:categoryId actions:actionsArray intentIdentifiers:@[] options:UNNotificationCategoryOptionNone];
```
Need to scope the category IDs in the Expo client (but not in standalone / detached apps). We'll need to apply the scoping on the server too.

ide commented on ios/Exponent/Versioned/Core/Api/EXNotifications.m:
```diff
@@ -92,41 +93,124 @@ - (instancetype)initWithExperienceId:(NSString *)experienceId kernelServiceDeleg
+  NSMutableArray<UNNotificationAction *> * actionsArray = [[NSMutableArray alloc] init];
+  
+  for(NSArray *action in actions) {
+    int optionsInt = [(NSNumber *)action[2] intValue];
```
The action parsing needs to be more robust to prevent runtime crashes and to communicate what each field means. Create a small struct that has five fields and write a "parser" that creates a struct from the array, and sets an `NSError **` if parsing fails. Then we can reject this promise if there is an error.

sjchmiela replied:
BUMP

ide commented on ios/Exponent/Versioned/Core/Api/EXNotifications.m:
```diff
@@ -92,41 +93,124 @@ - (instancetype)initWithExperienceId:(NSString *)experienceId kernelServiceDeleg
+  UNNotificationCategory * newCategory = [UNNotificationCategory categoryWithIdentifier:categoryId actions:actionsArray intentIdentifiers:@[] options:UNNotificationCategoryOptionNone];
+  
+  [[UNUserNotificationCenter currentNotificationCenter] getNotificationCategoriesWithCompletionHandler:^(NSSet * categories) {
+    NSMutableSet * categoriesMutable = [categories mutableCopy];
```
name this `newCategories` or something more descriptive that communicates the intent

ide commented on ios/Exponent/Versioned/Core/Api/EXNotifications.m:
```diff
@@ -92,41 +93,124 @@ - (instancetype)initWithExperienceId:(NSString *)experienceId kernelServiceDeleg
+    }
+    [categoriesMutable addObject:newCategory];
+    [[UNUserNotificationCenter currentNotificationCenter] setNotificationCategories:categoriesMutable];
+    resolve(@(YES));
```
`resolve(nil)` if the return value does not contain any information (0 bits)

ide commented on ios/Exponent/Versioned/Core/Api/EXNotifications.m:
```diff
@@ -92,41 +93,124 @@ - (instancetype)initWithExperienceId:(NSString *)experienceId kernelServiceDeleg
+  UNMutableNotificationContent* content = [self _localNotificationFromPayload:payload];
+  
+  NSDateComponents * date = [[NSDateComponents alloc] init];
+  NSArray * unites = @[@"day", @"month", @"year", @"weekday", @"quarter", @"leapMonth", @"nanosecond", @"era", @"weekdayOrdinal", @"weekOfMonth", @"weekOfYear", @"hour", @"second", @"minute", @"yearForWeekOfYear"];
```
unites -> units

Some of these seem not so useful or difficult to document. What is "era" or "leapMonth"? We need to support this on Android too.

ide commented on ios/Exponent/Versioned/Core/Api/EXNotifications.m:
```diff
@@ -92,41 +93,124 @@ - (instancetype)initWithExperienceId:(NSString *)experienceId kernelServiceDeleg
-
+  bool repeats = NO;
+  if (options[@"repeats"]) {
+    repeats = ([options[@"repeats"] isEqualToString:@"Yes"])? YES : NO;
```
I agree with @sjchmiela we probably want to preserve the existing `repeat` API for now, but if we keep this, the value should be a Boolean instead of the string "Yes"

sjchmiela replied:
bump, don't compare `options[@"repeats"]` with `"Yes"`, it should be an NSNumber

ide commented on packages/expo/src/Notifications.ts:
```diff
@@ -46,6 +46,9 @@ type Channel = {
 // Since that's not supported on iOS, we generate an unique string.
 type LocalNotificationId = string | number;
 
+const RED = 2;
```
Use TypeScript enums instead, using a plural name because the enum values are bit masks:

```
enum NotificationCategoryFlags {
  None = 0,
  AuthenticationRequired = 1,
  Destructive = 2,
  Foreground = 4,
};
```

The values of these options do not need to correspond to the runtime values that iOS uses because Android may offer different options and likely doesn't use the same enum values. There should be an Expo-to-iOS conversion step. 

Have you looked at the options Android provides?

ide commented on ios/Exponent/ExpoKit/ExpoKit.m:
```diff
@@ -112,9 +112,10 @@ - (void)application:(UIApplication *)application didFinishLaunchingWithOptions:(
-
+  
   // This is safe to call; if the app doesn't have permission to display user-facing notifications
   // then registering for a push token is a no-op
```
This comment applies to the remote notification manager code, keep the lines together

ide commented on ios/Exponent/ExpoKit/ExpoKit.m:
```diff
@@ -112,9 +112,10 @@ - (void)application:(UIApplication *)application didFinishLaunchingWithOptions:(
+  
   // This is safe to call; if the app doesn't have permission to display user-facing notifications
   // then registering for a push token is a no-op
+  [[EXUserNotificationManager sharedInstance] authorizeAndInit:launchOptions];
```
Is this safe to call? It looks like it will ask the user for permission to show notifications. It also needs to run on the notifications queue for iOS 10 & 11.

ide commented on ios/Exponent/Kernel/Services/EXSendNotificationParams.h:
```diff
@@ -0,0 +1,18 @@
+#import <Foundation/Foundation.h>
+
+@interface EXSendNotificationParams : NSObject
+@property (atomic, strong) NSString *experienceId;
```
Use nonatomic unless atomic is truly needed

ide commented on ios/Exponent/Kernel/Services/EXSendNotificationParams.h:
```diff
@@ -0,0 +1,18 @@
+@interface EXSendNotificationParams : NSObject
+@property (atomic, strong) NSString *experienceId;
+@property (atomic, strong) NSDictionary *dic;
+@property (atomic, strong) NSNumber *isRemote;
```
Just keep these as `(assign) BOOL`

ide commented on ios/Exponent/Kernel/Services/EXSendNotificationParams.m:
```diff
@@ -0,0 +1,31 @@
+
+@implementation EXSendNotificationParams
+
+- (instancetype)initWithExpId:(NSString *)expId
```
write out `initWithExperienceId`

ide commented on ios/Exponent/Kernel/Services/EXSendNotificationParams.m:
```diff
@@ -0,0 +1,31 @@
+@implementation EXSendNotificationParams
+
+- (instancetype)initWithExpId:(NSString *)expId
+   notificationBody: (NSDictionary *)dic
```
name this `body` instead of `dic`

ide commented on ios/Exponent/Kernel/Core/EXKernel.m:
```diff
@@ -139,10 +142,22 @@ - (void)sendUrl:(NSString *)urlString toAppRecord:(EXKernelAppRecord *)app
 
+- (void)flushPendingNotifications
+{
+  if (_pendingNotificationParams == nil) return;
```
Use curly braces consistently - 
```
if (...) {
  return;
}

[self sendNotification:...];
```

ide commented on ios/Exponent/Kernel/Core/EXKernel.m:
```diff
@@ -15,6 +15,7 @@
 #import <React/RCTEventDispatcher.h>
 #import <React/RCTModuleData.h>
 #import <React/RCTUtils.h>
+#import "EXSendNotificationParams.h"
```
Move this import with the other imports above, sorted alphabetically

ide reviewed (changes requested):
The pending notification flow looks complicated to me. It seems like the main use case is for when the user launches (cold start) the Expo client by tapping a notification -- we need to load Home, then send the notification to Home's JS, and then load the experience and give it the notification. If this is the use case that the pending notification solves, I'd like to simplify the code so there are fewer pieces that touch this notification or flush the queue so the notification isn't passed around as much and this code is less likely to break.

The queues are another area I think we need to refactor. In iOS 10 and 11, Apple said that UNNotificationCenter can run on any thread, but also that it is not thread-safe. In iOS 12, they changed that so UNNotificationCenter is thread-safe. I think a good approach would be to define a globally shared dispatch queue call `EXNotifications.notificationsQueue` that the notifications code uses. This would allow us to run off of the main thread while remaining thread-safe. (@sjchmiela do you have opinions on how to manage the concurrency here?)

Finally, I think it will be difficult (more difficult than many other APIs) to figure out the right API without implementing this for Android as well, and we probably should wait at least one SDK cycle before setting the expectation that we'll keep the API stable.

ide commented on ios/Exponent/Versioned/Core/Api/EXNotifications.m:
```diff
@@ -209,6 +228,10 @@ - (instancetype)initWithExperienceId:(NSString *)experienceId kernelServiceDeleg
 RCT_EXPORT_METHOD(cancelScheduledNotification:(NSString *)uniqueId)
 {
+  if (![EXEnvironment sharedEnvironment].isDetached) {
+    uniqueId = [self prefixWithExperience:uniqueId];
```
@sjchmiela in other places of the codebase do we use the word "scope" or "prefix" more?

sjchmiela replied:
I think we could follow naming of `scopedModules` here. An easy refactoring win would be to always pass `uniqueId` through some function named differently, eg. `[self categoryIdForUserId:uniqueId]` (I'm not really a fan of `UserId`, but you get the gist). This function could `if isDetached`.

ide commented on ios/Exponent/Versioned/Core/Api/EXNotifications.m:
```diff
@@ -292,4 +319,8 @@ - (UNMutableNotificationContent *)_localNotificationFromPayload:(NSDictionary *)
 }
 
+-(NSString *) prefixWithExperience:(NSString *) identifier {
+  return [NSString stringWithFormat:@"%@%@%@", self.expId, @"@@@", identifier];
```
Just use `:` instead of `@@@`. We don't allow `:` in usernames or project names (we validate that project slugs match a regex that doesn't include `:`).

sjchmiela commented on ios/Exponent/Kernel/Services/EXUserNotificationManager.m:
```diff
@@ -0,0 +1,91 @@
+  completionHandler(UNNotificationPresentationOptionAlert + UNNotificationPresentationOptionSound);
+}
+
+- (void)setNotificationDelegate
```
Name of this method isn't the best description of its content. Maybe we should just set the delegate in the initializer? I suppose this is how it's done in EXAnalytics — https://github.com/expo/expo/pull/2316/files#diff-0c6b818bdfc75ebed906c186926f43c4R102

sjchmiela replied:
bump

sjchmiela commented on ios/Exponent/Kernel/Services/Notifications/EXRemoteNotificationManager.m:
```diff
@@ -115,54 +100,57 @@ - (void)getExpoPushTokenForScopedModule:(id)scopedModule
-      }
+
+    [self _canRegisterForRemoteNotificationsWithHandler:^(BOOL can) {
+        if (!can) {
```
bump

sjchmiela commented on ios/Exponent/Kernel/Services/Notifications/EXRemoteNotificationManager.m:
```diff
@@ -208,26 +196,24 @@ - (void)_synchronizeCurrentAPNSToken:(NSData *)currentToken
-  UIUserNotificationSettings * settings = RCTSharedApplication().currentUserNotificationSettings;
-  return settings.types != UIUserNotificationTypeNone;
+  [[UNUserNotificationCenter currentNotificationCenter] getNotificationSettingsWithCompletionHandler:^(UNNotificationSettings * _Nonnull settings) {
+    handler(settings.authorizationStatus == UNAuthorizationStatusAuthorized);
```
bump

sjchmiela commented on ios/Exponent/Versioned/Core/Api/EXNotifications.m:
```diff
@@ -92,41 +94,140 @@ - (instancetype)initWithExperienceId:(NSString *)experienceId kernelServiceDeleg
   [EXUtil performSynchronouslyOnMainThread:^{
-    [RCTSharedApplication() presentLocalNotificationNow:notification];
+    UNNotificationRequest* request = [UNNotificationRequest
+                                      requestWithIdentifier:content.userInfo[@"id"] content:content trigger:nil];
```
Either make this method call in one line or go with `[UNNotificationRequest requestWithIdentifier:…NEWLINE`

sjchmiela commented on ios/Exponent/Versioned/Core/Api/EXNotifications.m:
```diff
@@ -92,41 +94,140 @@ - (instancetype)initWithExperienceId:(NSString *)experienceId kernelServiceDeleg
+  NSArray * units = @[@"day", @"month", @"year", @"weekday", @"quarter", @"leapMonth", @"nanosecond", @"era", @"weekdayOrdinal", @"weekOfMonth", @"weekOfYear", @"hour", @"second", @"minute", @"yearForWeekOfYear"];
+  for( NSString * unit in units) {
+    if (options[unit]) {
+      [date setValue:(NSNumber *)options[unit] forKey:unit];
```
This would look nicer extracted to another function

sjchmiela replied:
Please, extract it to another function

sjchmiela commented on ios/Exponent/Versioned/Core/Api/EXNotifications.m:
```diff
@@ -92,41 +94,140 @@ - (instancetype)initWithExperienceId:(NSString *)experienceId kernelServiceDeleg
+  }
+  UNCalendarNotificationTrigger* trigger = [UNCalendarNotificationTrigger triggerWithDateMatchingComponents:date repeats:repeats];
+  UNNotificationRequest* request = [UNNotificationRequest
+                                    requestWithIdentifier:content.userInfo[@"id"] content:content trigger:trigger];
```
wrong formatting

sjchmiela commented on ios/Exponent/Versioned/Core/Api/EXNotifications.m:
```diff
@@ -92,41 +94,140 @@ - (instancetype)initWithExperienceId:(NSString *)experienceId kernelServiceDeleg
+  
   [EXUtil performSynchronouslyOnMainThread:^{
-    [RCTSharedApplication() scheduleLocalNotification:notification];
+    int timeInterval = [((NSNumber *)options[@"time-interval"]) intValue];
```
options from JS should always be camelcased

sjchmiela commented on modules/expo-permissions/ios/EXPermissions/EXPermissionUserNotificationCenter.h:
```diff
@@ -0,0 +1,17 @@
+
+NS_ASSUME_NONNULL_BEGIN
+
+@protocol EXUserNotificationsPermissionsCenterInterface <NSObject>
```
Interfaces should live in `expo-permissions-interface`

sjchmiela commented on modules/expo-permissions/ios/EXPermissions/EXPermissionUserNotificationCenter.m:
```diff
@@ -0,0 +1,27 @@
+
+#import "EXPermissionUserNotificationCenter.h"
```
Do not use `import ""` in universal modules

sjchmiela commented on modules/expo-permissions/ios/EXPermissions/EXUserNotificationRequester.m:
```diff
@@ -0,0 +1,100 @@
+
+@implementation EXUserNotificationRequester
+
+static EXModuleRegistry * _moduleRegistry = nil;
```
Don't use `moduleRegistry` like that

Szymon20000 commented on apps/native-component-list/screens/UserNotifications.js:
```diff
@@ -0,0 +1,154 @@
+      },
+    };
+
+    await Notifications.setCategoryAsync('super-category', [action, textAction]);
```
setCategoryForIdAsync

sjchmiela commented on apps/native-component-list/screens/UserNotifications.js:
```diff
@@ -0,0 +1,154 @@
+  };
+
+  _waitTenSec = async () => {
+    this.notificationId = await Notifications.scheduleLocalNotificationWithTimeIntervalAsync(
```
`scheduleLocalNotificationWithTimeIntervalIOSAsync`

sjchmiela commented on ios/Exponent/Kernel/Core/EXKernel.m:
```diff
@@ -168,9 +185,15 @@ - (void)sendNotification:(NSDictionary *)notifBody
           fromBackground:(BOOL)isFromBackground
                 isRemote:(BOOL)isRemote
+                actionId: (NSString *)actionId
+                userText: (NSString *)userText
```
bump

sjchmiela commented on ios/Exponent/Kernel/Services/EXSendNotificationParams.h:
```diff
@@ -0,0 +1,24 @@
+//  Exponent
+//
+//  Created by smszymon on 21.09.2018.
+//  Copyright © 2018 650 Industries. All rights reserved.
```
bump

sjchmiela commented on ios/Exponent/Kernel/Services/EXSendNotificationParams.h:
```diff
@@ -0,0 +1,24 @@
+           isRemote: (NSNumber *) isRemote
+   isFromBackground: (NSNumber *)isFromBackground
+           actionId: (NSString *)actionId
+           userText: (NSString *)userText;
```
bump
also, copy and paste this whole function declaration — this will align `:`s

sjchmiela commented on ios/Exponent/Kernel/Services/EXSendNotificationParams.m:
```diff
@@ -0,0 +1,24 @@
+
+@implementation EXSendNotificationParams
+
+- (instancetype)initWithExperienceId:(NSString *)experienceId
```
Copy and paste this whole function implementation, this will align `:`s

sjchmiela commented on ios/Exponent/Versioned/Core/Api/EXCategoryAction.h:
```diff
@@ -0,0 +1,24 @@
+#import <Foundation/Foundation.h>
```
Maybe you could move EXNotifications and this class to a new folder `Notifications` under `Api` directory

sjchmiela commented on ios/Exponent/Versioned/Core/Api/EXCategoryAction.m:
```diff
@@ -0,0 +1,52 @@
+
+@implementation EXCategoryAction
+
++ (instancetype)parseFromArray: (NSArray*) array {
```
Do not parse `EXCategoryAction` from array. Pass a dictionary from JS and extract information from that dictionary, not using numbers as indexes, but string keys.

It would be much easier to read if such method would be written as follows:

```
- (UNNotificationAction *)notificationActionFromParams:(NSDictionary *)params
{
  if (params are of text input) {
    // configure and return UNTextInputNotificationAction
  }
  
  // configure and return UNNotificationAction
}
```

I think your approach is good and would suit well some long-lived objects (the subclassing hierarchy), but for such helper function I think this is an overkill

sjchmiela commented on ios/Exponent/Versioned/Core/Api/EXCategoryAction.m:
```diff
@@ -0,0 +1,52 @@
+//
```
Copyright comment

sjchmiela commented on ios/Exponent/Versioned/Core/Api/EXNotifications.m:
```diff
@@ -4,11 +4,14 @@
 
 #import <EXConstantsInterface/EXConstantsInterface.h>
+#import <UserNotifications/UserNotifications.h>
+#import "EXCategoryAction.h"
```
Let's move this import under `EXEnvironment`

sjchmiela commented on ios/Exponent/Versioned/Core/Api/EXNotifications.m:
```diff
@@ -92,41 +95,134 @@ - (instancetype)initWithExperienceId:(NSString *)experienceId kernelServiceDeleg
-  resolve(notification.userInfo[@"id"]);
+}
+
+RCT_EXPORT_METHOD(addCategory: (NSString *)categoryId
```
name of the method exported to the user should match name of the native method; append `Async`

sjchmiela commented on ios/Exponent/Versioned/Core/Api/EXNotifications.m:
```diff
@@ -92,41 +95,134 @@ - (instancetype)initWithExperienceId:(NSString *)experienceId kernelServiceDeleg
+  
+  [[EXUserNotificationCenter sharedInstance] getNotificationCategoriesWithCompletionHandler:^(NSSet * categories) {
+    NSMutableSet * newCategories = [categories mutableCopy];
+    for( UNNotificationCategory * category in newCategories) {
```
formatting

sjchmiela commented on ios/Exponent/Versioned/Core/Api/EXNotifications.m:
```diff
@@ -92,41 +95,134 @@ - (instancetype)initWithExperienceId:(NSString *)experienceId kernelServiceDeleg
-  UILocalNotification *notification = [self _localNotificationFromPayload:payload];
+  bool repeats = NO;
+  if (options[@"repeats"]) {
+    repeats = (BOOL)options[@"repeats"];
```
`options[@"repeats"]` will be either `nil` or an `NSNumber`. To coalesce this value to a bool, use `boolValue` method on NSNumber

sjchmiela commented on ios/Exponent/Versioned/Core/Api/EXNotifications.m:
```diff
@@ -178,29 +275,40 @@ - (instancetype)initWithExperienceId:(NSString *)experienceId kernelServiceDeleg
+}
 
-  return localNotification;
+-(NSString *) getScopedIdIfDetached:(NSString *) identifier {
```
formatting

sjchmiela commented on packages/expo/src/Notifications.ts:
```diff
@@ -42,10 +42,51 @@ type Channel = {
 // Since that's not supported on iOS, we generate an unique string.
 type LocalNotificationId = string | number;
 
+const FlagsEnum = Object.freeze({
```
Please export these constants from native side

sjchmiela commented on packages/expo/src/Notifications.ts:
```diff
@@ -244,13 +312,28 @@ export default {
 
+  async scheduleLocalNotificationWithMatchAsync(notification, options) {
+    if (Platform.OS === 'android') {
+       return new Promise((resolve,reject) => reject("This function currently works only on ios"));
```
You could just throw an Error
```
throw new Error("This function is supported only on iOS.")
```

sjchmiela commented on packages/expo/src/Notifications.ts:
```diff
@@ -244,13 +312,28 @@ export default {
     }
   },
 
+  async scheduleLocalNotificationWithMatchAsync(notification, options) {
```
TS typings

sjchmiela commented on packages/expo/src/Notifications.ts:
```diff
@@ -173,6 +214,33 @@ export default {
   },
 
+  // User passes set of actions titles.
+  putCategoryAsync(categoryId: string, actions: Array<ActionType>): Promise<string> {
```
This promise doesn't resolve to a string

sjchmiela commented on packages/expo/src/Notifications.ts:
```diff
@@ -42,10 +42,51 @@ type Channel = {
 type LocalNotificationId = string | number;
 
+const FlagsEnum = Object.freeze({
+  "None":0,
```
Why `""` around keys

sjchmiela reviewed (changes requested):
Some comments

Commit:
Updated notifications

Moved new permissions files to packages

fix

updated yarn.lock

fixes

Requested changes

fix  after rabase

Made requested changes

Renamed UserNotificationCenterProxy classes

Run `pod install`

ide commented on ios/Exponent/ExpoNotificationCenter/EXUserNotificationCenter.h:
```diff
@@ -0,0 +1,30 @@
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface EXUserNotificationCenter : NSObject
```
Add a comment explaining that this class provides thread-safety for UNUserNotificationCenter on iOS 10 and 11. We don't want to add more methods to this class unless UNUserNotificationCenter adds those methods, and can remove the class when we drop support for iOS 10 and 11.

ide commented on packages/expo-permissions/ios/EXPermissions/EXUserNotificationRequester.m:
```diff
@@ -0,0 +1,102 @@
+  __block BOOL allowsBadge;
+  __block EXPermissionStatus status;
+
+  [[EXUserNotificationRequester getCenterWithModuleRegistry:moduleRegistry] getNotificationSettingsWithCompletionHandler:^(UNNotificationSettings *settings) {
```
This looks like it could be non-thread-safe -- need to dispatch onto the notifications queue.

ide commented on packages/expo-permissions/ios/EXPermissions/EXUserNotificationRequester.m:
```diff
@@ -0,0 +1,102 @@
+
+    status = EXPermissionStatusUndetermined;
+
+    if (settings.authorizationStatus == UNAuthorizationStatusAuthorized) {
```
I believe we can treat Provisional as if the permission were granted, but need to test

ide replied:
When we get "provisional" permission (this means being able to send a notification without showing the permission dialog to the user), we'll get back UNAuthorizationStatusProvisional and should send back "granted" (and maybe `"provisional": true` in the dictionary).

ide commented on packages/expo-permissions/ios/EXPermissions/EXUserNotificationRequester.m:
```diff
@@ -0,0 +1,102 @@
+
+    if (settings.authorizationStatus == UNAuthorizationStatusAuthorized) {
+      status = EXPermissionStatusGranted;
+    } else {
```
Need to check for Denied explicitly

ide commented on packages/expo/src/Notifications.ts:
```diff
@@ -173,6 +208,11 @@ export default {
   },
 
+  // User passes set of actions titles.
+  putCategoryAsync(categoryId: string, actions: Array<ActionType>): Promise<void> {
```
Name this `createCategoryAsync` or `defineCategoryAsync`

ide commented:
Also can you run Xcode's formatter (https://freakycoder.com/xcode-tips-tricks-1-how-to-reformat-the-code-69fcfb400ab3) over the new files?

Szymon20000 commented on packages/expo/src/Notifications.ts:
```diff
@@ -244,13 +284,28 @@ export default {
       repeat?: 'minute' | 'hour' | 'day' | 'week' | 'month' | 'year';
       intervalMs?: number;
+      dateMatchingTriggerIOS?: DateMatchingTrigger;
+      intervalTriggerIOS?: IntervalTrigger;
```
I believe that intervalMs is something else because docs say "(Android only) intervalMs (optional) (number) -- Repeat interval in number of milliseconds" so probably user can pass time option different than intervalMs. I think that we can't easily provide that for iOS.

ide commented on packages/expo/src/Notifications.ts:
```diff
@@ -287,18 +342,42 @@ export default {
+      if (Platform.OS === "ios" ) {
+        warning(
+          options.repeat == undefined,
+          'This function is deprecated for ios with repeat option'
```
```suggestion
          'The "repeat" option for scheduled local notifications has been removed on iOS. Use "dateMatchingTriggerIOS" instead.'
```

sjchmiela committed:
Post-review fixes

ide commented on ios/Exponent/Kernel/Core/EXKernel.m:
```diff
@@ -215,34 +217,7 @@ - (void)_dispatchJSEvent:(NSString *)eventName body:(NSDictionary *)eventBody to
 
 - (NSDictionary *)initialAppPropsFromLaunchOptions:(NSDictionary *)launchOptions
```
We'll need to update the docs so that people don't expect the root component's `props.exp` to contain the launch notification. (This has come up in bug reports / issues a few times.)

sjchmiela replied:
I expect initial notification to appear in `props.exp`. 🙂 With `UNNotificationFramework` we can't use `launchOptions` to access the initial notification synchronously, but as soon as we set a delegate on `UNNotificationCenter.defaultCenter`, the delegate is called with `userNotificationCenter:didReceiveNotificationResponse:`. Our delegate is `EXUserNotificationManager`, which, if the kernel couldn't send the notification to a running application, saves the notification in `<experienceId, pendingNotification>` dictionary. When the app starts to run, `EXReactAppManager` asks `EXUserNotificationManager` for initial notification for given `experienceId`.

Notification could be lost if `EXReactAppManager` would ask `EXUserNotificationManager` for pending notification before `UNNotificationCenter` would notify delegate of the notification response. I can't say I'm 100% sure this will never happen, but as the Apple documentation states:

> Notifications that cause your app to be launched are delivered shortly after these methods ([`application:didFinishLaunchingWithOptions:`]) finish executing.

ide replied:
OK, thanks for the explanation.

> Notification could be lost if EXReactAppManager would ask EXUserNotificationManager for pending notification before UNNotificationCenter would notify delegate of the notification response.

In this scenario -- if it happens -- would we send the notification to the app when it subscribes to new notification events? We may still need to save up pending notifications after an app has launched to make sure we handle this scenario:

1. App launches. `EXReactAppManager` finds no pending notification from `EXUserNotificationManager`.
2. iOS invokes `userNotificationCenter:didReceiveNotificationResponse:`. This could happen either because (a) the app was launched from a notification and iOS delivered it late or (b) the app received a push notification after the app launched.
3. App JS subscribes to notification events.
4. How do we get the notification from step 2?

ide replied:
Reading the docs more, it makes me think that:

1. For now, we should keep the behavior in this PR. Apple's docs make it sound like they invoke the delegate method very soon after the app launches since they want the notification delegate to be set up before `application:didFinishLaunchingWithOptions:` returns. This will likely behave correctly for Expo's current API of providing the launch notification in `props.exp`.

2. In the future (this PR certainly has enough changes!) I would like to feel even more confident in our notification API and defend against changes in Apple's behavior by changing Expo's API to be closer to Apple's. There are two potential race condition bugs I see. One is that the launch notification gets sent to the notification delegate after the JS app launches (too late to pass it in through `props.exp`). The second is that after the native app launches, we may receive a notification response before the JS app launches if the timing is just right. In this case, we'd possibly (a) have two pending notifications if the app was launched from a different notification or (b) incorrectly think the notification response was used to launch the app and store it as a pending notification that gets passed to the JS app through `props.exp`.

To address (2), I think we'd need to provide an API similar to Apple's. Specifically, instead of passing notifications through `props.exp` we'd always pass notifications through the event emitter. (Apple removed `launchOptions` and now always uses the delegate method => Expo removes `props.exp.notification` and now always uses the event listener.)

Also similar to Apple's API, we'd require Expo developers to subscribe to the notification listener in JS before a cutoff point that we choose (for example, before `Expo.registerComponent` is called?)  -- if they do this, we guarantee that the notification used to launch the app (if the app was launched from a notification) will get sent to the listener. We can implement this by buffering up **all** pending notifications the native app receives before the cutoff point (ex: `Expo.registerComponent()`) and then replaying all of them by emitting events. This way the app is guaranteed to receive the notification that launched the app and will also receive all notifications that were delivered after app launched -- perfect delivery.

ide commented on packages/expo-permissions/ios/EXPermissions/EXUserNotificationRequester.m:
```diff
@@ -0,0 +1,103 @@
+  return [moduleRegistry getModuleImplementingProtocol:@protocol(EXUserNotificationCenterProxyInterface)];
+}
+
++ (NSDictionary *)permissionsWithModuleRegistry:(EXModuleRegistry *)moduleRegistry
```
Since this method is blocking, we may want to add:
```
dispatch_assert_queue_not(dispatch_get_main_queue())
```

ide commented on packages/expo-permissions/ios/EXPermissions/EXUserNotificationRequester.m:
```diff
@@ -0,0 +1,103 @@
+  UNAuthorizationOptions options = UNAuthorizationOptionAlert + UNAuthorizationOptionSound + UNAuthorizationOptionBadge;
+  id<EXUserNotificationCenterProxyInterface> notificationCenter = [EXUserNotificationRequester getCenterWithModuleRegistry:_moduleRegistry];
+  [notificationCenter requestAuthorizationWithOptions:options completionHandler:^(BOOL granted, NSError * _Nullable error) {
+    dispatch_async(dispatch_get_main_queue(), ^{
```
Move this off of the main thread. `-[_consumeResolverWithCurrentPermissions]` is blocking with the semaphore code, so we want to be especially careful not to run it on the main thread (see the comment about the assertion earlier).

ide commented on packages/expo-permissions/ios/EXPermissions/EXUserNotificationRequester.m:
```diff
@@ -0,0 +1,103 @@
+  id<EXUserNotificationCenterProxyInterface> notificationCenter = [EXUserNotificationRequester getCenterWithModuleRegistry:_moduleRegistry];
+  [notificationCenter requestAuthorizationWithOptions:options completionHandler:^(BOOL granted, NSError * _Nullable error) {
+    dispatch_async(dispatch_get_main_queue(), ^{
+      [weakSelf _consumeResolverWithCurrentPermissions];
```
Instead of storing `resolve` and `reject` as instance properties, I think it would be better to have a separate property for whether we are currently requesting permissions. That way, `resolve` and `reject` can't be called from other places in this class:

 ```objc
strongSelf = weakSelf
if (!strongSelf) {
  return;
}

strongSelf->_isRequestingPermissions = NO;

if (error) {  // should handle errors too
  reject(...);
  return;
}

resolve([strongSelf class] permissionsWithModuleRegistry:_moduleRegistry])
if (strongSelf->_delegate) {
  ...
}
```

ide replied:
The difference is that `resolve` and `reject` are kept within the scope of the code that provided them (the caller of `-[requestPermissionsWithResolver:(EXPromiseResolveBlock)resolve rejecter:(EXPromiseRejectBlock)reject]`) and  one call to `requestPermissionsWithResolver` cannot access the `resolve` and `reject` blocks of another call to `requestPermissionsWithResolver`.

I think it'd also achieve the same thing to add `resolve` as a parameter of `_consumeResolverWithCurrentPermissions`. 

If this class becomes more complex, it will be harder to ensure that we call the correct `resolve` and `reject` blocks if they are stored as instance properties. Sometimes using the instance properties is necessary but in this case I think having fewer properties would be better. The point is to not expose `resolve` and `reject` as instance properties -- the state we need to store in an instance property is only the fact that a concurrent call to `requestPermissionsWithResolver` is happening, and we don't need the power to resolve/reject that concurrent call to be exposed to other calls.

ide commented on packages/expo/src/Notifications.ts:
```diff
@@ -173,6 +184,15 @@ export default {
   },
 
+  // User passes set of actions titles.
+  createCategoryIOSAsync(categoryId: string, actions: Array<ActionType>): Promise<void> {
```
Use `[]` syntax for simple arrays:
```suggestion
  createCategoryIOSAsync(categoryId: string, actions: ActionType[]): Promise<void> {
```

sjchmiela commented on ios/Exponent/Kernel/Core/EXKernel.m:
```diff
@@ -215,34 +217,7 @@ - (void)_dispatchJSEvent:(NSString *)eventName body:(NSDictionary *)eventBody to
-    @"remote": @(isRemote),
-    @"data": notifBody,
-  };
+  return nil;
```
This path of code could be in fact removed altogether, as it was only used to add the initial notification. The only place where `initialProps` are stored and used is:

https://github.com/expo/expo/blob/ad80e41a81f3df7b24b9205b7bad55823cf4b3b3/ios/Exponent/Kernel/ReactAppManager/EXReactAppManager.m#L583-L585

So it's safe to return `nil` from this function until it's removed. 🙂

ide replied:
OK, can you please add a nullable annotation to the return type then?

sjchmiela commented on ios/Exponent/Versioned/Core/Api/EXNotifications.m:
```diff
@@ -176,31 +182,137 @@ - (instancetype)initWithExperienceId:(NSString *)experienceId kernelServiceDeleg
+  return [calendar components:unitFlags fromDate:triggerDate];
+}
+
+- (void)mutateDateComponents:(NSDateComponents *)dateComponents toRepeatEvery:(NSCalendarUnit)timePeriod
```
This code can't handle a use case when one wants to schedule a notification that will trigger eg. every minute, but starting next week (in such case it will start triggering this week).

sjchmiela replied:
It's not possible to use a subclass of `UNCalendarNotificationTrigger`, adding a notification request with a custom trigger fails with
```
Error Domain=NSCocoaErrorDomain Code=4097 "connection to service named com.apple.usernotifications.usernotificationservice"
```

So, to sum up we can't replicate scheduling a notification that will repeat every minute starting 3 hours from now.

We could either remove support for `repeat` option on iOS or only in such case use `UILocalNotification` which is internally transformed to `UNNotificationRequest` by iOS (with a trigger of a non-public class `UNLegacyNotificationTrigger`).

What do you think? cc @ide @tsapeta

sjchmiela replied:
I've added `ExpoNotifications.legacyScheduleLocalRepeatingNotification` exported native method which uses deprecated `UILocalNotification` to schedule a repeating notification with first fire date in the future. Using such configuration warns that this will be removed in next SDK release. The change has been neatly committed in https://github.com/expo/expo/pull/2316/commits/9fc0237ad1a2406af140fd794461374cb3790b43, before which I've implemented throwing in such case (`(time && repeat)`, https://github.com/expo/expo/pull/2316/commits/89068f90caf6d7684a8ebd678c0f46553faf9edd), so it's easily revertible.

sjchmiela replied:
I think along with deprecation of (`time && repeat`) we should add ability to schedule a notification date-components-way, but let's do this in a separate PR.

sjchmiela committed:
Minor JS fixes

sjchmiela committed:
Do not remember pending notifications in Expo Client (behavior matching previous one)

sjchmiela committed:
Prevent blocking main queue at all times by EXUserNotificationRequester

sjchmiela committed:
Do not scope categories identifiers in ejected applications

sjchmiela commented on packages/expo-permissions/ios/EXPermissions/EXLocationRequester.m:
```diff
@@ -83,8 +84,11 @@ - (void)requestPermissionsWithResolver:(EXPromiseResolveBlock)resolve rejecter:(
-    _locMgr = [[CLLocationManager alloc] init];
-    _locMgr.delegate = self;
+    __weak typeof(self) weakSelf = self;
+    [EXUtilities performSynchronouslyOnMainThread:^{
```
TODO: Add explanatory comment mentioning https://stackoverflow.com/a/12395192/1123156

sjchmiela replied:
Also, see whether this fixed the “Location permission issue” when asking for multiple permissions in one call.

ide replied:
Could you please split this out into a different commit (feel free to commit it directly to master)? This will make it easier to run git bisect or revert commits, which is generally useful and especially helpful for this massive PR.

sjchmiela replied:
Done in https://github.com/expo/expo/commit/304fe560500b662be53be2c1d5a06445ad9d3702. 🙂

ide commented on ios/Exponent/Kernel/Services/Notifications/EXRemoteNotificationManager.m:
```diff
@@ -208,26 +196,27 @@ - (void)_synchronizeCurrentAPNSToken:(NSData *)currentToken
-  return settings.types != UIUserNotificationTypeNone;
+  [[EXUserNotificationCenter sharedInstance] getNotificationSettingsWithCompletionHandler:^(UNNotificationSettings * _Nonnull settings) {
+    bool canRegister = settings.authorizationStatus == UNAuthorizationStatusAuthorized;
+    if (@available(iOS 12,  *)) {
```
```suggestion
    if (@available(iOS 12, *)) {
```

ide commented on ios/Exponent/Versioned/Core/Api/EXNotifications.m:
```diff
@@ -92,63 +94,67 @@ - (instancetype)initWithExperienceId:(NSString *)experienceId kernelServiceDeleg
+                                                                        trigger:nil];
+
+  [[EXUserNotificationCenter sharedInstance] addNotificationRequest:request withCompletionHandler:^(NSError * _Nullable error) {
+    if (error != nil) {
```
```suggestion
    if (error) {
```

ide commented on packages/expo-permissions/ios/EXPermissions/EXUserNotificationRequester.m:
```diff
@@ -0,0 +1,107 @@
+
+  UNAuthorizationOptions options = UNAuthorizationOptionAlert + UNAuthorizationOptionSound + UNAuthorizationOptionBadge;
+  id<EXUserNotificationCenterProxyInterface> notificationCenter = [EXUserNotificationRequester getCenterWithModuleRegistry:_moduleRegistry];
+  [notificationCenter requestAuthorizationWithOptions:options completionHandler:^(BOOL granted, NSError * _Nullable error) {
```
We need to handle this error somehow. At minimum, we should log it, but I think it'd be better to reject the promise or send the error information back to JS somehow.

ide commented on ios/Exponent/Kernel/Services/EXUserNotificationManager.m:
```diff
@@ -0,0 +1,58 @@
+
+# pragma mark - UNUserNotificationCenterDelegate
+
+- (void)userNotificationCenter:(UNUserNotificationCenter *)center didReceiveNotificationResponse:(UNNotificationResponse *)response withCompletionHandler:(void (^)(void))completionHandler
```
It looks like this method is called also when the user dismisses a notification from the iOS notification center UI. We probably don't want to send an event to JS in that case.

sjchmiela replied:
Have you experienced it or read in the docs? From my experience and understanding we would needed to add a custom category with [`UNNotificationCategoryOptionCustomDismissAction `](https://developer.apple.com/documentation/usernotifications/unnotificationcategoryoptions/unnotificationcategoryoptioncustomdismissaction?language=objc) option for this method to be called when the user dismisses the notification, but I'll reassure that during QA of this PR.

sjchmiela replied:
Confirmed! This method won't be called when a user dismisses notification with current configuration.

ide commented on ios/Exponent/Kernel/Services/EXUserNotificationManager.m:
```diff
@@ -0,0 +1,58 @@
+  // With UIUserNotifications framework, notifications were only shown while the app wasn't active.
+  // Let's stick to this behavior.
+  if ([UIApplication sharedApplication].applicationState != UIApplicationStateActive) {
+    completionHandler(UNNotificationPresentationOptionAlert + UNNotificationPresentationOptionSound);
```
Need `UNNotificationPresentationOptionBadge` too?

sjchmiela replied:
Thanks for catching that!

ide commented on ios/Exponent/Kernel/Services/EXUserNotificationManager.m:
```diff
@@ -0,0 +1,58 @@
+    _pendingNotifications[pendingNotification.experienceId] = pendingNotification;
+  }
+
+  completionHandler(UNNotificationPresentationOptionNone);
```
I think we want to set the badge number even when delivering the notification: UNNotificationPresentationOptionBadge

sjchmiela replied:
On `master` the badge isn't set if the notification is delivered while the app is active. 🙂

sjchmiela commented on ios/Exponent/Kernel/Services/EXUserNotificationManager.m:
```diff
@@ -0,0 +1,58 @@
+- (void)userNotificationCenter:(UNUserNotificationCenter *)center didReceiveNotificationResponse:(UNNotificationResponse *)response withCompletionHandler:(void (^)(void))completionHandler
+{
+  EXPendingNotification *pendingNotification = [[EXPendingNotification alloc] initWithNotificationResponse:response];
+  if (pendingNotification && ![[EXKernel sharedInstance] sendNotification:pendingNotification] && [EXEnvironment sharedEnvironment].isDetached) {
```
It is always set, but not always returned (`initWithNotificationResponse :` may return `nil` if `EXPendingNotification` created from the response would be invalid). I'll change it to some more explicit `isSchedulable` method call, not a nullable initializer.

sjchmiela replied:
Sorry, my whole comment is outdated. 😞 Accepting your suggestion.

sjchmiela committed:
Apply review comments

sjchmiela committed:
Remove obsolete if condition

sjchmiela committed:
Do not store resolver and rejecter in EXUserNotificationRequester

sjchmiela committed:
Update Pods

sjchmiela committed:
Add support for deleting custom notification categories

sjchmiela committed:
Pass notification to a remembered experience when tapped a notification in backgrounded Expo Client

sjchmiela committed:
Update jest-expo mocks

sjchmiela committed:
Ensure permissions consuming happens on permissions queue in EXRemoteNotificationRequester

sjchmiela committed:
Do not scope notification identifiers

sjchmiela committed:
Change cancelScheduledNotification into cancelScheduledNotificationAsync returning native Promise

sjchmiela committed:
Deny scheduleNotificationAsync with both time and repeat set on iOS

sjchmiela committed:
Allow scheduleNotificationAsync with both time and repeat set on iOS using deprecated UILocalNotification

sjchmiela committed:
Add more notification testing buttons to NCL

sjchmiela committed:
Export EXUserNotificationManager service as kernel service

sjchmiela committed:
Add EXNotificationsIdentifiersManager protocol capable of scoping and unscoping identifiers

sjchmiela committed:
Require both RemoteNotificationManager and UserNotificationManager in EXNotifications module

sjchmiela committed:
Scope and unscope action identifiers by EXUserNotificationManager

sjchmiela committed:
Update pods
